crypto/x509/x509_vfy.c

```diff
@@ -267,9 +267,12 @@ int X509_verify_cert(X509_STORE_CTX *ctx)
     int ret;
 
     if (ctx->cert == NULL) {
-        ERR_raise(ERR_LIB_X509, X509_R_NO_CERT_SET_FOR_US_TO_VERIFY);
-        ctx->error = X509_V_ERR_INVALID_CALL;
-        return -1;
+        if (sk_X509_num(ctx->untrusted) < 1) {
+            ERR_raise(ERR_LIB_X509, X509_R_NO_CERT_SET_FOR_US_TO_VERIFY);
+            ctx->error = X509_V_ERR_INVALID_CALL;
+            return -1;
+        }
+        ctx->cert = sk_X509_value(ctx->untrusted, 0);
```

Comment:
I am not sure why this change (to implicitly use the first untrusted certificate as verify target if cert is NULL) is needed.

Reply:
It is not needed, but very natural! 
See comment just above and https://github.com/openssl/openssl/pull/13735#discussion_r565654494 and the subsequent ones there.

Reply:
It's just a scope creep for the PR. :grin: Not saying that you have to do a separate one for it, but anyway.

Reply:
True - glad that you are graceful on this :smile:
I've just added the new feature to the PR intro text.

Reply:
This change turns what used to be an error into doing something.  Sigh.

Reply:
> This change turns what used to be an error into doing something. Sigh.

@richsalz, most people would welcome such a neat new feature, but you sigh on it. Sigh.

Reply:
Burying a new feature in a PR that's about documentation (hint, you can edit the title), and *changing the API* is a concern for 3.0 which promises most things will work if you just recompile.  Did you miss that?

*I AM AGAINST CHANGES TO THE API FOR 3.0*

I think the OMC needs to consider this API change, like they have for most other API changes.

Reply:
I have to agree with rich that this change is unnecessary and does not look to me as valuable as to fight for it. So better would be to either drop it or spawn it to a new PR.

Reply:
> I AM AGAINST CHANGES TO THE API FOR 3.0

This is not an API _change_. It's an addition; nothing can break on including this.

> I think the OMC needs to consider this API change, like they have for most other API changes.

Recently we still had numerous (mostly larger) extensions to the API going in, with no OMC decision on it. 

Anyway I've separated that commit, to follow @t8m's request such that this PR can be finished more quickly.

Reply:
This is perhaps my fault.  I had dimly remembered that this was *existing* behaviour when reviewing the documentation, but perhaps that was something vaguely similar in some other part of the library.  I had not intended to suggest a change in behaviour.

Upon closer inspection, indeed there was no implicit use of the first cert as the cert to verify, so we should probably not change that. Sorry about the confusion...

Reply:
> It's an addition; nothing can break on including this.

Sure it can.  Programs which used to get an error (e.g., couldn't read the cert because of permissions) will now get a new result.

>Recently we still had numerous (mostly larger) extensions to the API going in, with no OMC decision on it.

That's unfortunate, if accurate. Doesn't necessarily make it right.  We've also seen breaking changes that *were* put to the OMC to discuss and vote.